### PR TITLE
Prevent toast autohiding if focusing or hovering

### LINKED ISSUE #32820: Toasts should not autohide when hovered or when focus is inside
For accessibility and usability reasons, if an autohiding toast is hovered with the mouse, or when `:focus-within`, it should prevent the autohiding / reset the delay (so when not hovered anymore / when focus moves out again, the delay kicks in again and eventually the toast does hide).

Due to `:focus-within`, probably can only target v5 - unless extra work is done in JS to check `document.activeElement` and to verify if it's a child element of the toast.

### PR DESCRIPTION
**Pull request by RyanBerliner**
resolves #32820

TODO:

- [ ] Land it after some of the refactoring PRs so that we don't need to adapt bundlewatch

### CODE REVIEW

File: js/src/toast.js [hunk 1]
```diff
@@ -157,6 +159,10 @@ class Toast extends BaseComponent {
     }
 
     EventHandler.off(this._element, EVENT_CLICK_DISMISS)
+    EventHandler.off(this._element, EVENT_MOUSEOVER)
```
Comment (GeoSot) [inline js/src/toast.js:162]: You can replace all of them, using just `EventHandler.off(this._element, EVENT_KEY)` ;)

File: js/src/toast.js [hunk 2]
```diff
@@ -64,6 +68,8 @@ class Toast extends BaseComponent {
 
     this._config = this._getConfig(config)
     this._timeout = null
+    this._activeMouseInteraction = false
+    this._activeKeyboardInteraction = false
```
Comment (GeoSot) [inline js/src/toast.js:72]: Does it matter if it is mouse or keyboard interaction instead of 'interaction'? Do you have something in mind?

File: js/src/toast.js [hunk 3]
```diff
@@ -176,8 +182,53 @@ class Toast extends BaseComponent {
...
+    if (this._activeMouseInteraction || this._activeKeyboardInteraction) {
+      return
+    }
+
+    this._timeout = setTimeout(() => {
+      this.hide()
+    }, this._config.delay)
+  }
+
+  _onInteraction(event, interacting) {
```
Comment (GeoSot) [inline js/src/toast.js:199]: maybe `isInteracting` would be more descriptive, as it is boolean

File: js/src/toast.js [hunk 4]
```diff
@@ -176,8 +182,53 @@ class Toast extends BaseComponent {
...
+      default:
+        break
+    }
+
+    if (interacting) {
+      this._clearTimeout()
+      return
+    }
+
+    const nextElement = event.relatedTarget
```
Comment (GeoSot) [inline js/src/toast.js:218]: this is to avoid multiple triggering events?

File: js/src/toast.js (hunk 2 above)
Comment (RyanBerliner) [reply js/src/toast.js:72]: Mouse interactions and keyboard interactions need to be tracked separately. For example, if the user has both their cursor, and focus inside of the toast, then their cursor leaves the toast, we need a mechanism in place that still knows about the focus inside the toast element.

For concrete examples of the scenario I'm referring to, take a look at the tests
- "should not auto hide if focus leaves but cursor remains inside"
- "should not auto hide if cursor leaves but focus remains inside"

If those scenarios are something that should be supported, and you can still think of a different way to accomplish the desired behavior, LMK! If we want just 1 property, we couple make it an object like this:
```javascript
this._activeInteraction = {}

// Later in the file, for example
this._activeInteraction.mouse = true|false
```
Although, this is just a surface change, there is nothing fundamentally different about it.

File: js/src/toast.js (hunk 4 above)
Comment (RyanBerliner) [reply js/src/toast.js:214]: This is an optimization, if this conditional is removed everything would still work fine. 

At this point in the event handler, we know that the interaction is a "leaving" one. Either `mouseout` or `focusout`. While we could just call `this._maybeScheduleHide()`, why bother if we **know** any timer it might set would be cancelled almost immediately? By checking whether or not we are about to receive an "entering" interaction (`mouseover` or `focusin`) we can save a round trip of creating, and then immediately clearing a timer.

Comment (GeoSot) [reply js/src/toast.js:214]: The `nextElement` I mean , but I guess is to  check the [previous ](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/relatedTarget#javascript)target . Right?

File: js/src/toast.js (hunk 2 above)
Comment (GeoSot) [reply js/src/toast.js:72]: I am not sure of keyboard functionality. How many times do we achieve to catch a toast by tabbing into it?
/Cc @patrickhlauke may be the right person to help

Also, as `_activeMouseInteraction` and `_activeKeyboardInteraction` are also booleans, it seems right to prefix them with `has` (_hasActiveMouseInteraction)

File: js/src/toast.js (hunk 4 above)
Comment (RyanBerliner) [reply js/src/toast.js:214]: I think we are talking about the same thing, I might just be unclear with my explanation. Let me try again - `nextElement` is the most meaningful term to describe `event.relatedTarget` in this case. At this point in the conditional the event is a "leaving" event. Either `mouseout` or `focusout`. The `relatedTarget` of a leaving event is the element that the users interaction is moving to _next_, just like the mdn link you referenced outlines. So in the case of a `mouseout`, `nextElement` assigned the element that the cursor will move to next, and in the case of `focusout`, `nextElement` is assigned to the element the the focus will move to next.

So, the conditional referencing the `nextElement` says "if the user is just moving to another element in the toast, don't even bother trying to schedule a hide".

Comment (GeoSot) [reply js/src/toast.js:214]: I get it. thank you.

File: js/src/toast.js (hunk 2 above)
Comment (RyanBerliner) [reply js/src/toast.js:72]: The original issue states both cursor and focus interaction should prevent the toast from hiding. If the team decides otherwise then this can be simplified like you where originally suggesting, though I'd recommend against it.

File: js/tests/unit/toast.spec.js [hunk 5]
```diff
@@ -210,6 +210,182 @@ describe('Toast', () => {
...
+        })
+
+        const mouseOverEvent = createEvent('mouseover')
+        toastEl.dispatchEvent(mouseOverEvent)
+      }, toast._config.delay / 2)
+
+      toast.show()
+    })
+
+    it('should not auto hide if focus leaves but cursor remains inside', done => {
```
Comment (patrickhlauke) [inline js/tests/unit/toast.spec.js:316]: sorry, took me far too long to parse that "cursor" here refers to the mouse pointer (as, when using screen readers, there is a separate concept of reading cursor which can be different from focus). suggest just changing the wording here from "cursor" to "mouse pointer")

File: js/tests/unit/toast.spec.js [hunk 6]
```diff
@@ -210,6 +210,182 @@ describe('Toast', () => {
...
+        })
+
+        const mouseOverEvent = createEvent('mouseover')
+        toastEl.dispatchEvent(mouseOverEvent)
+      }, toast._config.delay / 2)
+
+      toast.show()
+    })
+
+    it('should not auto hide if cursor leaves but focus remains inside', done => {
```
Comment (patrickhlauke) [inline js/tests/unit/toast.spec.js:353]: same as above...suggest changing "cursor" to "mouse pointer"

**Review by patrickhlauke — CHANGES_REQUESTED** [review]

File: js/src/toast.js (hunk 2 above)
Comment (patrickhlauke) [reply js/src/toast.js:72]: yes, makes sense to check for both mouse over and for actual keyboard focus being inside the toast

**Review by patrickhlauke — APPROVED** [review]

File: js/src/toast.js (hunk 2 above)
Comment (GeoSot) [inline js/src/toast.js:72]: please, try to follow boolean variables naming convention, putting the prefix 'has'

**Review by GeoSot — CHANGES_REQUESTED** [review]

File: js/src/toast.js [hunk 7]
```diff
@@ -174,8 +176,53 @@ class Toast extends BaseComponent {
...
+        break
+    }
+
+    if (isInteracting) {
+      this._clearTimeout()
+      return
+    }
+
+    const nextElement = event.relatedTarget
+    if (this._element === nextElement || this._element.contains(nextElement)) {
```
Comment (GeoSot) [inline js/src/toast.js:213]: @RyanBerliner, does #33310  changes the needs for these checks?

File: js/tests/unit/toast.spec.js [hunk 8]
```diff
@@ -210,6 +210,182 @@ describe('Toast', () => {
...
+        })
+
+        toastEl.addEventListener('focusin', () => {
+          const mouseOutEvent = createEvent('mouseout')
+          toastEl.dispatchEvent(mouseOutEvent)
+        })
+
+        toastEl.addEventListener('mouseout', () => {
+          expect(toast._timeout).toBeNull()
+          done()
```
Comment (GeoSot) [inline js/tests/unit/toast.spec.js:380]: Can add `expect(toast._hasKeyboardInteraction).toEqual(true)` (not really important)

File: js/tests/unit/toast.spec.js [hunk 9]
```diff
@@ -210,6 +210,182 @@ describe('Toast', () => {
...
+        })
+
+        toastEl.addEventListener('focusin', () => {
+          const outsideFocusable = document.getElementById('outside-focusable')
+          outsideFocusable.focus()
+        })
+
+        toastEl.addEventListener('focusout', () => {
+          expect(toast._timeout).toBeNull()
+          done()
```
Comment (GeoSot) [inline js/tests/unit/toast.spec.js:343]: Can add `expect(toast._hasMouseInteraction).toEqual(true)` (not really important)

File: js/src/toast.js (hunk 7 above)
Comment (RyanBerliner) [reply js/src/toast.js:213]: If we were _only_ concerned with mouse pointer interactions then yes, that event handler PR would allow us to use `mouseenter` and `mouseleave` and omit this check. However, since we are using `focusin` and `focusout` here as well, it's most clear IMO to use the equivalent mouse events and share all the same handler logic. If you have a compelling case otherwise I'm not opposed to changing this.

Comment (GeoSot) [reply js/src/toast.js:213]: @rohit2sharma95 @alpa any opinions?

Comment (alpadev) [reply js/src/toast.js:213]: I'm fine with either way. For mouse events it might be not needed anymore but like Ryan mentioned it makes sense for focus events and it's not like it really bloats the code IMO.

File: js/tests/unit/toast.spec.js (hunk 8 above)
Comment (RyanBerliner) [reply js/tests/unit/toast.spec.js:380]: I'm going to omit this for simplicities sake.

File: js/tests/unit/toast.spec.js (hunk 9 above)
Comment (RyanBerliner) [reply js/tests/unit/toast.spec.js:343]: I'm going to omit this for simplicities sake.

**Review by GeoSot — APPROVED** [review]

**Comment by GeoSot** [comment]
@RyanBerliner  please rebase your branch.

**Review by GeoSot — APPROVED** [review]

**Review by GeoSot — APPROVED** [review]
